File: 128LongestConsecutive/longestConsecutive.py
class Solution:
  def longestConsecutive(self, nums: List[int]) -> int:
    st = set(nums)
    res = 0
    len_st = len(st)
    for n in st:
      if(n - 1 in st):
        continue
      length = 1
      while(n + length in st):
        length = length + 1
      res = max(length, res)
      if res * 2 > len_st:
        break 
    return res

File: 128LongestConsecutive/longestConsecutive.cpp
#include <unordered_set>
#include <vector>

using namespace std;

/**
 * 在数组内查找元素，用哈希表把查找时间复杂度降到O(1)
 */
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> u_set(nums.begin(), nums.end());
        int res = 0;
        for (int i : u_set) {
            if (u_set.find(i - 1) != u_set.end()) {
                continue;
            }
            int len = 1;
            while (u_set.find(i + len) != u_set.end()) {
                len++;
            }
            res = max(res, len);
        }
        return res;
    }
};
